### PR TITLE
Enhance processing of primary commodity groups

### PR DESCRIPTION
**Pull request by olejandro**
This pull request treats creation of commodity groups and assigment of default primary commodity groups to processes.
Commodity groups can have one or more commodities in them. In case of a single commodity, the commodity group is the commodity itself.
Veda creates commodity groups based on the commodity type and whether the commodity is in input or in output. E.g., a process `PRC1`, producing commodities `COM1` and `COM2` of type `NRG` would result in the creation of `PRC1_NRGO` commodity group, containing `COM1` and `COM2`; `COM1` and `COM2` would be valid commodity groups for `PRC1`, as well.
Assuming that `COM3` is the sole input to `PRC1` of type `NRG`, this pull request also addresses the case where `NRGI` is specified as a default commodity group for `PRC1`. I.e. `PRC1_NRGI` commodity group would be created with `COM3` as the only member.

### CODE REVIEW

**Comment by olejandro** [comment]
⬆️ 2% on TIM and 3-11% ⬆️ on demos

**Comment by olejandro** [comment]
Thanks @siddharth-krishna! I've added a description to explain what's going on. 😃 
Let me rebase this onto main instead of merging.

**Comment by olejandro** [comment]
Ah, so the checks fail because of the longer run time...

**Comment by siddharth-krishna** [comment]
Ah yes sorry perhaps my criterion of 10% increase in runtime is too strict? Should we up it to 50%?

**Comment by olejandro** [comment]
Hmmm... I wonder whether it makes sense to increase the runtime regression treshhold to e.g. 50% or 100%, since we are not yet at 100% ground truth match?

**Comment by siddharth-krishna** [comment]
Good point. I propose making it 100% (or a factor of 2). We're also not really doing a robust timing here -- ideally runtimes would be measured on repeated runs and on a dedicated machine without other tasks running.

The factor of increase of runtime of this PR are:
```
DemoS_001 1.0420168067226891
DemoS_002 1.1466666666666667
DemoS_003 1.1797752808988764
DemoS_004 1.1770833333333333
DemoS_005 1.2226148409893993
DemoS_006 1.2046204620462047
DemoS_007 1.3089430894308944
DemoS_008 1.5973025048169553
DemoS_009 1.575809199318569
DemoS_010 1.5711892797319935
DemoS_011 1.5687919463087248
DemoS_012 1.5408320493066254
Ireland 1.3036823425022184
```

File: times_reader/transforms.py [hunk 1]
```diff
@@ -780,75 +780,151 @@ def extract_commodity_groups(
...
+                if default_set:
+                    break
+                for cset in csets_ordered_for_pcg:
+                    df = comm_groups.loc[
+                        (comm_groups["Region"] == region)
+                        & (comm_groups["TechName"] == process)
+                        & (comm_groups["IO"] == io)
+                        & (comm_groups["Csets"] == cset)
+                    ]
+                    if df.shape[0]:
```
Comment (samwebster) [inline times_reader/transforms.py:896]: df.empty might be clearer?

File: times_reader/transforms.py [hunk 2]
```diff
@@ -859,12 +935,76 @@ def extract_commodity_groups(
...
+    def expand_pcg_from_suffix(df):
+        """
+        Return the name of a default primary commodity group based on suffix and process name
+        """
+
+        # Veda determines default PCG based on this order and presence of OUT/IN commodity
+        csets_ordered_for_pcg = ["DEM", "MAT", "NRG", "ENV", "FIN"]
+        default_pcg_suffixes = [
+            cset + io for cset in csets_ordered_for_pcg for io in ["I", "O"]
+        ]
```
Comment (samwebster) [inline times_reader/transforms.py:961]: Worth factoring out? I'm not sure

**Comment by samwebster** [comment]
Thanks so much for doing this. The code I wrote that's been replaced was the worst in the project and just there to give me some confidence we had most of the data we needed. This is a big step forward that the increases in accuracy don't do justice

**Review by samwebster — APPROVED** [review]

**Comment by olejandro** [comment]
> Thanks so much for doing this. The code I wrote that's been replaced was the worst in the project and just there to give me some confidence we had most of the data we needed. This is a big step forward that the increases in accuracy don't do justice to.

Thanks, Sam! This was a good opportunity for me to finally find out how exactly commodity groups are generated for TIMES. It is one of those things that very few people look into; most just use. I bet this won't be the final version in any case.

**Pull request merged**